### PR TITLE
Add "custom link types", hyperlinks that require an outfit attribute to use

### PR DESCRIPTION
**Pull request by FranchuFranchu**
**Feature:** This PR implements "Custom links", hyperlinks that require an outfit attribute to use

## Feature Details
This can be useful to plugin developers, who might have an idea for a species that uses a special type of hyperdrive that can travel across their systems. The player could then purchase the outfit from them (or get it from a mission), which would allow him/her to go through their systems.

This can be kind of achieved right now by creating links with an event, but it's not very good because it means other ships that in the lore aren't meant to have that hyperdrive could travel through the new link too.

## UI Screenshots
Image
Here we can see that the player has a hyperdrive with the "kappa drive" attribute, which allows them to travel through the "kappa" custom links (colored yellow). The other links are greyed out because the player lacks the attribute needed (but they can be made invisible in the data files).

## Usage Examples
The plugin developer can define a "linktype" which specifies the requirement, and an outfit that has the required attribute.

    linktype "Kappa"
        color 0.5 0.5 0 1 # Color when the player is e
        "far color" 0.5 0.5 0 0.5 # Color when the player is far away from the link
        "unusable color" 0.5 0.5 0 0.3 # Color when the player lacks the attribute
        requires "Kappa Drive" # Requirement

    outfit "Kappa Hyperdrive"
        category "Systems"
        cost 50000
        thumbnail "outfit/hyperdrive"
        "mass" 20
        "Kappa Drive" 1 # The important attribute
        "outfit space" -20
        "jump speed" .2
        "jump fuel" 100
        "hyperdrive" 1

    system "Kappa Star"
        pos -100404 -100356
        government Independent
        habitable 425.92
        belt 1786
        link "Sigma Star"
        link "Theta Star"
        "custom link" "target system" Kappa # The special link

## Testing Done
[This temporary folder](https://github.com/FranchuFranchu/endless-sky/tree/master/testing%20plugin) can be used as a plugin to test the custom links. Obviously, i'll move the folder to another repo when this leaves draft status.

## Performance Impact
N/A

## Why this is a draft
The main reason of why this is a draft is that i've started working on this a year ago, which means that I haven't updated the rest of the code since 9a31318a7e26241f9fe868fd36e5f1db7a3cdcd3. I also want to hear your opinions about it before merging it manually.

### CODE REVIEW

**Comment by Terin** [comment]
Would be a very cool feature to have access to, yeah. Back when I started Endless Sky and before I got a jump drive, and saw inaccessible systems on the map, I thought either some mission would manually create hyperlinks to them, or an outfit would. Something like this can be a great way to flavor sections of the galaxy, like keystones do for the Ember Waste right now, especially if you could provide custom sounds/effects to them.

**Comment by ravenshining** [comment]
For links to be truly invisible, I think you'd need to include alpha in the colour?  Otherwise you might see a black line cutting across a galaxy image.

**Comment by FranchuFranchu** [comment]
> For links to be truly invisible, I think you'd need to include alpha in the colour? Otherwise you might see a black line cutting across a galaxy image.
@ravenshining 

The fourth value for the color in the usage example is the alpha value

**Comment by ravenshining** [comment]
Oh, I see that now.  Sorry, at first it looked like there were only three numbers there; I think my brain filled a decimal in between 0 and 1 when it's a space.

**Comment by EjoThims** [comment]
I still think this would be a great tool to have, even if we don't immediately include it in anything vanilla.

**Comment by Zitchas** [comment]
I agree. Could be very useful for stuff I have planned for later on in the Remnant campaign, too. And, well, reducing the number events playing havoc with the galaxy map is probably a good thing, too. (using events was my current plan...)

But yeah, having links that are flat out invisible except when a specific outfit is equipped would be *very* useful.

**Comment by EjoThims** [comment]
Gonna bump this one again too.

@FranchuFranchu any chance you find this ready to leave draft?

**Comment by EjoThims** [comment]
This would still be great to have.

Requesting a review, but with @FranchuFranchu not responding to the above king, I think this probably is ready for "Oprhaned" tag?

**Comment by EjoThims** [comment]
Marauder description changes are out of scope though, so if this revived or picked up by another, those should be split off, if continued at all.

**Comment by FranchuFranchu** [comment]
It seems like this PR got a lot of attention recently. I'll start working on it again, and update it for the current codebase. @EjoThims

**Comment by FranchuFranchu** [comment]
The latest commit I pushed to my branch is ready for testing. Here is a plugin you can use to test custom link types out. Escorts work well.

There are a few issues, for example, it doesn't take into account the edge case where the ship has enough fuel for a normal hyperspace trip, but the custom hyperdrive requires more fuel and can't do the same trip.

[testplugin.zip](https://github.com/endless-sky/endless-sky/files/9448285/testplugin.zip)

File: source/CustomLink.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
+/* CustomLink.cpp
+Copyright (c) 2014 by Michael Zahniser
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
```
Comment (tibetiroka) [inline source/CustomLink.cpp:9]:
```suggestion
/* CustomLink.cpp
Copyright (c) 2022 by Michael Zahniser

Endless Sky is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later version.

Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.
*/
```
This change is not technically necessary, but this spacing is used for the copyright header in general.

File: source/CustomLink.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/* CustomLink.h
+Copyright (c) 2014 by Michael Zahniser
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
```
Comment (tibetiroka) [inline source/CustomLink.h:9]:
```suggestion
/* CustomLink.h
Copyright (c) 2022 by Michael Zahniser

Endless Sky is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later version.

Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.
*/
```
I believe you can put your own name in the copyright header, but you can keep it as-is if you want to.

**Review by tibetiroka — CHANGES_REQUESTED** [review]
This is not a proper code review; I only checked the comments for grammar-related issues. Seems pretty good for the most part.

File: source/CustomLink.h (hunk 2 above)
Comment (Zitchas) [reply source/CustomLink.h:9]: My general perspective is that if you made the file, you can (and perhaps even should) put your name in the header; but you don't have to.

File: source/CustomLink.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
+/* CustomLink.cpp
+Copyright (c) 2014 by Michael Zahniser
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/CustomLink.cpp:8]:
```suggestion
Copyright (c) 2022 by Michael Zahniser

Endless Sky is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later version.

Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

File: source/CustomLink.h (hunk 2 above)
Comment (Hurleveur) [reply source/CustomLink.h:9]: I find it really weird to copyright for smbd else, maybe we should not write a name at all if copyright is denied?

File: source/CustomLink.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		bool CanTravel(const Ship &ship) const;
+		bool CanTravel(const Outfit &outfit) const;
+
+		const Color &GetColorFor(const Ship &ship, bool isClose) const;
+
+	private:
+		bool doHyperdriveMovement = false;
+
+		// Link color when it is far away from the player
+		Color farColor = Color(0.5f, 1.f);
```
Comment (warp-core) [inline source/CustomLink.h:64]: These `Color`s should probably be pointers.
Define the default colour in `interfaces.txt` and have that be used if a cusomt colour isn't provided in the custom link type definition.

File: source/CustomLink.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		// Color when it is close to the player
+		Color closeColor  = Color(1.f, 1.f);
+
+		// Same as above, but used when the player can't travel through the links. Can be used to create "secret" links.
+		Color unusableFarColor = Color(0.5f, 0.f);
+		Color unusableCloseColor = Color(1.f, 0.f);
+
+		// Outfit attribute required to travel through this link
+		std::string requirement;
```
Comment (warp-core) [inline source/CustomLink.h:73]: This should be an `std::set`, and, ideally, you would support needing all of a list of attributes or any one attribute from a list (or some combination of the two).

**Comment by Amazinite** [comment]
Going through PRs older than three months to find PRs that could be closed.

Marking this PR as orphaned and closing it as there hasn't been any activity from OP in four months, and it's also been marked as a draft for the entire time that it has been opened. Anyone interested in the feature can reopen a new PR with a more finalized version of the changes here, or OP can reopen the PR once it is out of the draft phase.

**Comment by Hurleveur** [comment]
Well also to note that the recent conditional link PR allows this and much more.

**Comment by Amazinite** [comment]
I will note that it's not exactly the same. While with conditional links, you could make links appear and disappear depending on outfits the player has, that condition actually changes the link for every ship, not just those ships that have that outfit. This PR allows per-ship travel customization.
